### DISCUSSION TITLE
Help generating projects with `lccgen`

### DISCUSSION THREAD

**Discussion by SadE54** [open-license-manager/licensecc#204] [General]
I'm trying to use licgen to create a new project , but I'm not sure how to use it 
If I'm first trying that:
lccgen project init -n POM
nothing is happening (no text , even with verbose option)

Do I have to provide a rsa private key (link to .rsa files ? or the keys content in the command line ?)

lccgen project init --primary-key pom_private_key.rsa --public-key pom_public_key.rsa --project-name POM

return nothing :-/

**Comment by GitHubUserC** [comment]
I have this problem too, can anyone answer?

**Comment by siddjellali** [comment]
same here ! Project is still live ?

**Comment by emgstabilis** [comment]
[edited because formatting was garbled by github]

I am not sure about the latest version but here are my notes when I
evaluated open-license-manager. I don't guarantee that it will work for you
because I did not follow recent developments and I don't know the internal
details of open-license-manager.

The documentation of open-license-manager
<https://open-license-manager.github.io/licensecc/usage/issue-licenses.html>
suggests to have the executable lcc in your path and organize the project
as follows:

```
projects
└── DEFAULT       #(your project name)
    ├── include
    │   └── licensecc
    │       └── DEFAULT
    │           ├── licensecc_properties.h
    │           └── public_key.h
    ├── licenses
    │   └── test.lic
    └── private_key.rsa
```

Here, DEFAULT is the project name and would be replaced by MYPROJECT, for
example.

The idea is that each project has its settings and licenses. In addition to
DEFAULT (which is used for internal tests of open-license-manager), a new
project can be added by reconfiguring cmake with the following additional
variable -DLCC_PROJECT_NAME=xxx.

When I did so for  MYPROJECT , I ended up with the following directory
structure:

```
.
├── DEFAULT
│   ├── include
│   │   └── licensecc
│   │       └── DEFAULT
│   │           ├── licensecc_properties.h
│   │           └── public_key.h
│   └── private_key.rsa
└── MYPROJECT
    ├── include
    │   └── licensecc
    │       └── MYPROJECT
    │           ├── licensecc_properties.h
    │           └── public_key.h
    └── private_key.rsa
```

Also, lccgen can be used to generate the project (in fact this is how the
cmake configuration does it). When I tested this, the same directory
structure was generated, but for some reason the file licensecc_properties.h
is not created.

From the above structure we see that the build process generated both the
public and the private key, by using openssl.

For a start, it can be sufficient to remove the public and private key when
rebuilding the project, and this will regenerate both keys.

After compiling the library (cmake --build . --target install --config
Release), the structure of the install directory is as follows:

```
.
├── bin
│   ├── lccgen
│   └── MYPROJECT
│       └── lccinspector
├── include
│   └── licensecc
│       ├── datatypes.h
│       ├── licensecc.h
│       └── MYPROJECT
│           ├── licensecc_properties.h
│           └── public_key.h
└── lib
    ├── cmake
    │   ├── lccgen
    │   │   ├── lccgen-config.cmake
    │   │   ├── license_generator.cmake
    │   │   └── license_generator-release.cmake
    │   └── licensecc
    │       ├── licensecc-config.cmake
    │       └── licensecc-config-version.cmake
    └── licensecc
        └── MYPROJECT
            ├── cmake
            │   ├── licensecc.cmake
            │   └── licensecc-release.cmake
            └── liblicensecc_static.a
```

There is a global lccgen executable, which is common for all projects, and
then a lccinspector specific for each project. I suppose that this is
because the hardware identifier strategy is specific for each project, and
other variables as well.

Also, the include directory from each project's directory is copied as to
the installation include directory, and a static library is generated for
each project. Finally, some .cmake files are added for each project,
probably to facilitate integration if you know how to use it.

**Comment by jslyadneva** [comment]
This video may help https://youtu.be/BHb-p-4aaZE?si=_UehJQVSmzQA6fwl

**Comment by gcontini** [comment]
The original discussion is in [issue 139](https://github.com/open-license-manager/licensecc/issues/139).

When built as a submodule with the default configuration  `licensecc` generates a `projects` folder structure in your <<CMAKE_SOURCE_DIRECTORY>> containing one initialized project. I recommend you start experiment issue license with what has been prepared.

* here is how to [issue licenses](https://open-license-manager.github.io/licensecc/v2.1.0/usage/issue-licenses.html)

-------------------
If you do want to move that structure outside the source code you can either:
  - move the `projects` folder 
  - run `lccgen project initialize -p <<my-project-name>> -t <<templates_dir>> --verbose`
 
*templates_dir*: is the directory where public.key.inja  is `licensecc/src/templates/`

Then when you configure your project with `cmake` remember to set: `-DLCC_PROJECT_NAME <<my-project-name>>  -D LCC_PROJECTS_BASE_DIR <<projects_dir>>`

*projects_dir*: is the parent directory where you generated your project. It should look like this (in this case the project name is "POM"):

```
projects
└── POM
    ├── include
    │   └── licensecc
    │       └── POM
    │           └── public_key.h
    └── private_key.rsa
```
